src/App.tsx

```diff
@@ -1,14 +1,105 @@
 /* eslint-disable max-len */
-import React from 'react';
+/* eslint-disable no-console */
+import React, {
+  useCallback, useEffect, useMemo, useState,
+} from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
+import debounce from 'lodash/debounce';
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { Select } from './types/Select';
+
+export const filterTodos = (s: Select, q: string, t: Todo[]): Todo[] => {
+  const lowerAppliedQuery = q.toLowerCase();
+
+  return t
+    .filter(td => {
+      switch (s) {
+        case Select.COMPLETED:
+          return td.completed;
+        case Select.ACTIVE:
+          return !td.completed;
+        default:
+          return true;
+      }
+    })
+    .filter(td => td.title.toLowerCase().includes(lowerAppliedQuery));
+};
+
+export const getTodoById = (id: number, todoArr: Todo[]): Todo | null => (
+  todoArr.find(todo => todo.id === id) || null
+);
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodoId, setSelectedTodoId] = useState(0);
+  const [select, setSelect] = useState<Select>(Select.ALL);
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
+  const [isLoading, setIsLoading] = useState(true);
+
+  const handleTodosLoading = useCallback(() => {
+    setIsLoading(true);
+
+    const loadingTodos = async () => {
+      try {
+        const todosFromServer = await getTodos();
+
+        setTodos(todosFromServer);
+        setIsLoading(false);
+      } catch (error) {
+        console.log('you catched some error');
+      }
+    };
+
+    loadingTodos();
+  }, []);
+
+  useEffect(handleTodosLoading, []);
+
+  const visibleTodos = useMemo(() => (
+    filterTodos(select, appliedQuery, todos)
+  ), [appliedQuery, select, todos]);
+
+  const selectedTodo = useMemo(() => (
+    getTodoById(selectedTodoId, todos)
+  ), [selectedTodoId]);
+
+  const applyQuery = useCallback(debounce((value) => {
+    setIsLoading(false);
+
+    return setAppliedQuery(value);
+  }, 700), []);
+
+  const handleSelectedIdChange = useCallback((todoId) => {
+    setSelectedTodoId(todoId);
+  }, []);
+
+  const handleCloseClick = useCallback(() => {
+    setSelectedTodoId(0);
+  }, []);
+
+  const handleSelectChange = useCallback((event) => {
+    setSelect(event.target.value);
+  }, []);
+
+  const handleQueryChange = useCallback((event) => {
+    setIsLoading(true);
+    setQuery(event.target.value);
+    applyQuery(event.target.value);
+  }, []);
+
+  const handleInputReset = useCallback(() => {
+    setQuery('');
+    setAppliedQuery('');
+  }, []);
+
```

Comment:
so much wrappers 😶‍🌫️

Reply:
@ArturErmolenko I use debounce to show search results after a delay, and I need to use setQuery to control the input, and setAppliedQuery to update the list after the delay. I thought about moving the query and setQuery to the component, but it wouldn't decrease the amount of code

Is it possible to optimize it, or did you mean something else?

Reply:
I believe you could pass setState functions without wrappers , but its not critical in this case